### PR TITLE
[DeepSpeed] decouple `DeepSpeedConfigHF` from `Trainer`

### LINKED ISSUE #11954: Uncoupling ZeRO-3 weak ref bridge b/w Trainer and modeling_utils
https://github.com/huggingface/transformers/blob/fd6204b2a70d100800cb259a7fbddfc812631ed3/src/transformers/modeling_utils.py#L1168

So I briefly looked through the ZeRO-3 integration into the `Trainer` and this approach of creating a "bridge" between the `Trainer` and `modeling_utils` via a weak ref is neat.

However, what if I wanted to use ZeRO-3 with HF models outside the scope of the `Trainer` and `TrainingArguments`? Seems like I cannot at the moment.

There seems to be 4 places in `modeling_utils` where `is_deepspeed_zero3_enabled` is called, of which the only one that seems to be heavily tied to the custom `DeepSpeedConfigHF` class is the one referenced above.

Is it not possible to allow `Init()` args to come from the parent `from_pretrained()` method's `kwargs`?

@stas00 @sgugger

### PR DESCRIPTION
**Pull request by stas00**
As requested in https://github.com/huggingface/transformers/issues/11954 this PR 

* uncouples `DeepSpeedConfigHF` from `Trainer` so one can activate `zero.Init()` in `modeling_utils.py` and several other places w/o needing to rely on the HF `Trainer`.
* adds a new `LoggingLevel` ctx manager to `testing_utils.py`
* adds a new test testing  `DeepSpeedConfigHF` decoupled from the `Trainer`.
* starts a new doc

* well, through the PR things got renamed too, see the final diff for all the changes.

The plan is to merge this, then move all of the Deepspeed integration code into its own `src/transformers/deepspeed.py` and the docs into `docs/source/main_classes/deepspeed.rst`, since the integration has now outgrown the Trainer alone.

Fixes: https://github.com/huggingface/transformers/issues/11954

@sgugger

### CODE REVIEW

File: src/transformers/integrations.py [hunk 1]
```diff
@@ -284,26 +284,50 @@ def _set_if_auto(config, key, val):
 
 class DeepSpeedConfigHF:
```
Comment (stas00) [inline src/transformers/integrations.py:285]: As this is now becoming a public class, should we think of a better name? I feel it's a bit too cryptic at the moment.

The key is to differentiate from DeepSpeed's own classes with similar name and to make it clear that this is a transformers/HF version.

Comment (sgugger) [reply src/transformers/integrations.py:285]: I would just use `HfDeepSpeedConfig` as the general rule used in other names is to put the HF first with the flowercase `HfAPI` (HfAPI, HfFolder, HfArgumentParser for instance)

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -284,26 +284,55 @@ def _set_if_auto(config, key, val):
...
     """
-    This object contains Deepspeed configuration and can be quickly queried for things like zero stage.
+    This object contains a DeepSpeed configuration dict and can be quickly queried for things like zero stage.
```
Comment (sgugger) [inline src/transformers/integrations.py:287]:
```suggestion
    This object contains a DeepSpeed configuration dictionary and can be quickly queried for things like zero stage.
```
no abbreviation in the docs please :-)

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -284,26 +284,55 @@ def _set_if_auto(config, key, val):
...
 
-    We store a ``weakref`` of this object in the module's global to be able to access the config from areas where the
-    Trainer is not available (e.g. `from_pretrained` and `_get_resized_embeddings`).
+    A ``weakref`` of this object is stored in the module's globals to be able to access the config from areas where
+    things like the Trainer object is not available (e.g. `from_pretrained` and `_get_resized_embeddings`). Therefore
```
Comment (sgugger) [inline src/transformers/integrations.py:290]:
```suggestion
    things like the Trainer object is not available (e.g. ``from_pretrained`` and ``_get_resized_embeddings``). Therefore
```

File: src/transformers/integrations.py [hunk 4]
```diff
@@ -284,26 +284,55 @@ def _set_if_auto(config, key, val):
...
+    it's important that this object remains alive while the program is still running.
+
+    :class:`~transformers.Trainer` uses the ``DeepSpeedConfigHFTrainer`` subclass instead. That subclass has logic to
+    sync the configuration with values of :class:`~transformers.TrainingArguments` by replacing special placeholder
+    values: ``auto``. Without this special logic the DeepSpeed configuration is not modified in any way.
```
Comment (sgugger) [inline src/transformers/integrations.py:295]:
```suggestion
    values: ``"auto"``. Without this special logic the DeepSpeed configuration is not modified in any way.
```
I think it's a string no?

File: src/transformers/integrations.py [hunk 5]
```diff
@@ -284,26 +284,55 @@ def _set_if_auto(config, key, val):
...
-    same lifespan as the latter.
     """
 
-    def __init__(self, args):
-        self.config = None
-        self.stage = 0
-        self.offload = False
+    def __init__(self, config_file_or_dict):
+        # set global weakref object
+        deepspeed_config_hf_set(self)
```
Comment (sgugger) [inline src/transformers/integrations.py:304]: I would put the hf first here too, and begin with the action word: `set_hf_deepspeed_config`

File: src/transformers/integrations.py [hunk 6]
```diff
@@ -314,28 +343,23 @@ def is_zero3(self):
...
 
-    def config_process(self, args):
-        """
-        1. load json if the ``args.deepspeed`` is a path
-        2. replace any ``auto`` values in the config with the correct or recommended value
 
-        This is done as early as possible, before model is created, to allow ``is_deepspeed_zero3_enabled`` query and
-        getting to the early deepspeed config object during ``zero.Init()`` which needs whether fp16 is enabled, dtype,
-        etc.
+class DeepSpeedConfigHFTrainer(DeepSpeedConfigHF):
```
Comment (sgugger) [inline src/transformers/integrations.py:347]: Same here for the names: `HfTrainerDeepSpeedConfig`.

File: src/transformers/testing_utils.py [hunk 7]
```diff
@@ -648,6 +650,24 @@ def __repr__(self):
         return f"captured: {self.out}\n"
 
 
+@contextlib.contextmanager
+def LoggingLevel(level):
```
Comment (sgugger) [inline src/transformers/testing_utils.py:654]: Neat context manager!

File: src/transformers/training_args.py [hunk 8]
```diff
@@ -671,10 +671,12 @@ def __post_init__(self):
...
             # - must be run very last in arg parsing, since it will use a lot of these settings.
             # - must be run before the model is created.
-            from transformers.integrations import DeepSpeedConfigHF
+            from transformers.integrations import DeepSpeedConfigHFTrainer
 
-            # will be used later by the Trainer (leave self.deepspeed unmodified in case a user relies on it not to be modified)
-            self.deepspeed_config_hf = DeepSpeedConfigHF(self)
+            # will be used later by the Trainer
+            # note: leave self.deepspeed unmodified in case a user relies on it not to be modified)
+            self.deepspeed_config_hf = DeepSpeedConfigHFTrainer(self.deepspeed)
```
Comment (sgugger) [inline src/transformers/training_args.py:678]: Same here, would put the hf first.

**Review by sgugger — APPROVED** [review]
Great work! I like the idea of fully decoupling the integration from the `Trainer`!

File: src/transformers/testing_utils.py [hunk 9]
```diff
@@ -648,6 +650,24 @@ def __repr__(self):
...
+    """
+
+    This is a context manager to temporarily change transformers modules logging level to the desired value and have it
+    restored to the original setting at the end of the scope. For example ::
+
+       with LoggingLevel(logging.INFO): # run some core function that calls logger.info() and it should be logged
+
```
Comment (stas00) [inline src/transformers/testing_utils.py:661]: just noticed the styling formatter messed up:
```
with LoggingLevel(logging.INFO): 
    # run some core function that calls logger.info() and it should be logged
```

Comment (stas00) [reply src/transformers/testing_utils.py:661]: hmm, if I put code there it still messed it up:

```
+       with LoggingLevel(logging.INFO): AutoModel.from_pretrained("gpt2")
```

what am I doing wrong? I though `::` indicates that it's code block in a docstring.

Comment (sgugger) [reply src/transformers/testing_utils.py:661]: Not sure. Can we put the comment above the with?

Comment (stas00) [reply src/transformers/testing_utils.py:661]: I changed it to:
```

... Example ::

with LoggingLevel(logging.INFO): 
    AutoModel.from_pretrained("gpt2")
```

and it unwraps it.

Comment (stas00) [reply src/transformers/testing_utils.py:661]: It comes from:
````
python utils/style_doc.py src/transformers docs/source --max_len 119
```

Comment (stas00) [reply src/transformers/testing_utils.py:661]: I tried different indentation and adding paras after, to no avail.

Comment (stas00) [reply src/transformers/testing_utils.py:661]: ok, the :: must not be in the first para - and then it works

**Comment by stas00** [comment]
Thank you for the great feedback and suggestions, Sylvain.

On to the next PR to move the code and docs.

**Comment by dunalduck0** [comment]
@stas00 I have a question about this PR. My use case is HF trainer + DeepSpeed + hyperparameter search.

In **training_arg.py**: hf_deepspeed_config is first loaded from config file (e.g. zero3.config) as HfTrainerDeepSpeedConfig, and then is adjusted with TrainingArguments values. In particular, all "auto" in zero3.config would be resolved to actual integral or floating values.

            self.hf_deepspeed_config = HfTrainerDeepSpeedConfig(self.deepspeed)
            self.hf_deepspeed_config.trainer_config_process(self)

In **trainer.py -> _hp_search_setup()**: hf_deepspeed_config is reset by zero3.config to HfDeepSpeedConfig, different from HfTrainerDeepSpeedConfig, and thus "auto" values are remained unchanged as string type, not integral or floating type. 

            from transformers.deepspeed import HfDeepSpeedConfig
            self.args.hf_deepspeed_config = HfDeepSpeedConfig(self.args.deepspeed)

This didn't work for hyperparameter search because DS cannot be initialized with "auto" values. The error messages are attached below. I feel  _hp_search_setup() should do exact the same thing as in training_args.py to reset hf_deepspeed_config as HfTrainerDeepSpeedConfig and resolve all "auto" values, but I am not sure what was the reason you chose the other way in this PR.

  File "/home/meiyang/src/transformers_fork/src/transformers/integrations.py", line 164, in run_hp_search_optuna
    study.optimize(_objective, n_trials=n_trials, timeout=timeout, n_jobs=n_jobs)
  File "/home/meiyang/bin/miniconda3/envs/gptj/lib/python3.8/site-packages/optuna/study/study.py", line 400, in optimize
    _optimize(
  File "/home/meiyang/bin/miniconda3/envs/gptj/lib/python3.8/site-packages/optuna/study/_optimize.py", line 66, in _optimize
    _optimize_sequential(
  File "/home/meiyang/bin/miniconda3/envs/gptj/lib/python3.8/site-packages/optuna/study/_optimize.py", line 163, in _optimize_sequential
    trial = _run_trial(study, func, catch)
  File "/home/meiyang/bin/miniconda3/envs/gptj/lib/python3.8/site-packages/optuna/study/_optimize.py", line 264, in _run_trial
    raise func_err
  File "/home/meiyang/bin/miniconda3/envs/gptj/lib/python3.8/site-packages/optuna/study/_optimize.py", line 213, in _run_trial
    value_or_values = func(trial)
  File "/home/meiyang/src/transformers_fork/src/transformers/integrations.py", line 154, in _objective
    trainer.train(resume_from_checkpoint=checkpoint, trial=trial)
  File "/home/meiyang/src/transformers_fork/src/transformers/trainer.py", line 1155, in train
    self.model = self.call_model_init(trial)
  File "/home/meiyang/src/transformers_fork/src/transformers/trainer.py", line 1019, in call_model_init
    model = self.model_init()
  File "run_clm_local.py", line 289, in model_init
    pretrained_model = AutoModelForCausalLM.from_pretrained(
  File "/home/meiyang/src/transformers_fork/src/transformers/models/auto/auto_factory.py", line 447, in from_pretrained
    return model_class.from_pretrained(pretrained_model_name_or_path, *model_args, config=config, **kwargs)
  File "/home/meiyang/src/transformers_fork/src/transformers/modeling_utils.py", line 1488, in from_pretrained
    with deepspeed.zero.Init(config_dict_or_path=deepspeed_config()):
  File "/home/meiyang/src/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 461, in __init__
    _ds_config = DeepSpeedConfig(config_dict_or_path,
  File "/home/meiyang/src/deepspeed/deepspeed/runtime/config.py", line 873, in __init__
    self._configure_train_batch_size()
  File "/home/meiyang/src/deepspeed/deepspeed/runtime/config.py", line 1050, in _configure_train_batch_size
    self._batch_assertion()
  **File "/home/meiyang/src/deepspeed/deepspeed/runtime/config.py", line 986, in _batch_assertion
    train_batch > 0
TypeError: '>' not supported between instances of 'str' and 'int'**

**Comment by stas00** [comment]
Honestly, I have never used `_hp_search_setup()` and have no idea what it does, so it's very possible the DS integration doesn't support it at the moment as indicated by your report. 

Do you want to try and make it work and make a PR? 

I'm not exactly sure what you mean by:

> but I am not sure what was the reason you chose the other way in this PR.

but perhaps it'd be much simpler for you to code what you think it should be and then we can look together at what you meant.

How does that sound?

**Comment by dunalduck0** [comment]
Sure. I am working on it. After replacing to HFTrainerDeepSpeedConfig, I was able to pass DS initialization. But there are other issues for model init. Not sure if they’re DS related or not. Will continue looking tomorrow.

**Comment by stas00** [comment]
Thank you for working on this, @dunalduck0 - I trust that you will figure it out.

While you work on this please log all the steps so that we can reproduce the process and we will need to create tests to verify the workings of this path based on your logs as we currently don't have any tests exercising this path with deepspeed.

**Comment by dunalduck0** [comment]
Hi stas00, I wanted to update this thread a bit. To my question on Mar 3rd, I think the change below is good enough to make sure DeepSpeed configuration is loaded properly.  . 
image

But goal to use hyperparameter search feature in Transformers with DeepSpeed ([discussion thread](https://discuss.huggingface.co/t/using-hyperparameter-search-in-trainer/785/10)) is still blocking, because the difficulty of reuse an "used" DeepSpeed engine, similar to the problem you addressed last month: https://github.com/microsoft/DeepSpeed/issues/1748

The difference here is I don't need to save/load_checkpoint. For hyperparameters search to work, both trainer and DeepSpeed engine need to be reused in multiple trial trainings, to discover the optimal settings. I was managed to run trial 1 (the 1st time the DeepSpeed engine is used in training) and then trial 2 failed inside of DeepSpeed with "index out of range" error. It looks like some partition went wrong, but I am not sure. I wonder if you have reached any good solution with DeepSpeed team on this type of issues?

**Comment by stas00** [comment]
Some improvements have been made recently in Deepspeed, but last I tried to re-use the engine it still didn't work in all situations. You're of course welcome to open and issue at Deepspeed and ask for a better engine reuse. If you don't ask other priorities will take over.

But I'm sure we can find a workaround with what we have.

Bottom line, I need to study how this feature works, write a simple example that deploys this feature, then add DS support, then turn it into a test.

Thank you for the links with the usage examples. 

I'm currently very busy with the BigScience 176B model training launch, but hopefully next week I should have some time to tinker with this. Unless of course, you or someone else beats me to it ;)

**Comment by stas00** [comment]
@dunalduck0, my apologies for taking forever to attend to this feature request. 

Please try this PR https://github.com/huggingface/transformers/pull/16740 and let me know if it resolves your issue.

I only added a basic test, so if you're doing something specific that happens not to work please let me know and I will extend the test to include it.

**Pull request merged**